File: apps/android/app/src/main/kotlin/sh/nikhil/conduit/ui/AgentAvatar.kt
package sh.nikhil.conduit.ui

import android.annotation.SuppressLint
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.Code
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Compose mirror of `apps/ios/Sources/Views/AgentAvatar.swift`.
 *
 * Small circular avatar for an agent (claude, codex, hermes, pi,
 * opencode). Used in any place that lists or picks agents — the
 * `AgentPickerSheet` rows, the thread switcher row list, and the
 * `SessionInfoScreen` hero. Not used inside the chat composer or the
 * header pill — those are already tinted via
 * [ConduitTheme.accent].
 *
 * Renders a single-letter monogram on a filled disc using
 * [ConduitTheme.accentStrong]. Falling back to a letter (rather
 * than a logo) means we don't ship third-party brand marks in the
 * APK and the avatar works for any agent the harness exposes.
 */
@Composable
fun AgentAvatar(
    assistant: String,
    modifier: Modifier = Modifier,
    size: Dp = 24.dp,
) {
    val fill = agentAccentStrong(assistant)
    val onAccent = ConduitTheme.textOnAccent()
    val logoRes = agentLogoRes(assistant)
    val glyph = agentGlyph(assistant)
    val monogram = monogramFor(assistant)
    val label = assistant.replaceFirstChar { it.uppercaseChar() }

    Box(
        modifier = modifier
            .size(size)
            .clip(CircleShape)
            // Brand logos sit on a white disc (they're designed for a
            // light background — otherwise the black Codex knot is
            // invisible on the dark sheet); glyph/monogram fallbacks sit
            // on the accent disc.
            .then(if (logoRes == null) Modifier.background(fill) else Modifier.background(Color.White))
            .border(0.5.dp, onAccent.copy(alpha = 0.15f), CircleShape)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        when {
            // Divided by the optical-size correction (device feedback:
            // gemini/opencode read visibly smaller than claude/codex at the
            // same padding, since the source art isn't uniformly cropped)
            // so every agent occupies the same visual weight.
            logoRes != null -> Image(
                painter = painterResource(logoRes),
                contentDescription = null,
                contentScale = ContentScale.Fit,
                modifier = Modifier.fillMaxSize().padding(size * 0.16f / agentGlyphRenderScale(assistant)),
            )
            glyph != null -> Icon(
                // Claude / Codex get a distinctive brand glyph; other
                // agents keep the monogram.
                imageVector = glyph,
                contentDescription = null,
                tint = onAccent,
                modifier = Modifier.size(size * 0.5f),
            )
            else -> Text(
                text = monogram,
                color = onAccent,
                style = TextStyle(
                    fontFamily = FontFamily.SansSerif,
                    fontWeight = FontWeight.ExtraBold,
                    fontSize = (size.value * 0.5f).sp,
                ),
            )
        }
    }
}

/**
 * Bare tinted per-agent glyph -- no background disc/tile, just the real
 * brand mark (Claude starburst / Codex knot / opencode mark), or the icon
 * / monogram letter fallback when no mark is bundled, tinted with the
 * agent's theme color. This is the ConduitUI list-row idiom ("rows lead
 * with a bare tinted symbol, not a filled tile" -- see [ConduitNavRow] /
 * iOS `ConduitListRow`). Use this instead of [AgentAvatar] in any
 * row-leading position; reserve [AgentAvatar]'s filled disc for picker /
 * hero contexts that want a heavier visual anchor. Mirror of iOS
 * `AgentGlyph` (`apps/ios/Sources/Shared/AgentAvatar.swift`).
 */
@Composable
fun AgentGlyph(
    assistant: String,
    modifier: Modifier = Modifier,
    size: Dp = 20.dp,
) {
    val tint = agentAccent(assistant)
    val logoRes = agentLogoRes(assistant)
    val glyph = agentGlyph(assistant)
    val label = assistant.replaceFirstChar { it.uppercaseChar() }

    Box(
        modifier = modifier
            .size(size)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        when {
            logoRes != null -> Icon(
                // Real brand mark, tinted -- Icon's default ColorFilter
                // (BlendMode.SrcIn) replaces the source color entirely
                // using only the drawable's alpha, so this tints cleanly
                // even though the source PNG isn't a pure single color.
                // Scaled by the same optical-size correction as
                // [AgentAvatar] so gemini/opencode don't read smaller than
                // claude/codex; stays well inside the outer `size` Box, so
                // this can't overflow/clip.
                painter = painterResource(logoRes),
                contentDescription = null,
                tint = tint,
                modifier = Modifier.size(size * 0.62f * agentGlyphRenderScale(assistant)),
            )
            glyph != null -> Icon(
                imageVector = glyph,
                contentDescription = null,
                tint = tint,
                modifier = Modifier.size(size * 0.62f),
            )
            else -> Text(
                text = monogramFor(assistant),
                color = tint,
                style = TextStyle(
                    fontFamily = FontFamily.SansSerif,
                    fontWeight = FontWeight.ExtraBold,
                    fontSize = (size.value * 0.58f).sp,
                ),
            )
        }
    }
}

/**
 * Resolves the bundled brand-logo drawable for an agent, if the app owner
 * has supplied the official artwork (`claude_mark` / `codex_mark`). Looked
 * up by name at runtime so a missing drawable degrades to [agentGlyph] /
 * [monogramFor] rather than failing the build — we don't bundle the
 * artwork here; it's added under the trademark attribution in the
 * Licenses screen. Returns null when absent or for agents without a logo.
 */
@SuppressLint("DiscouragedApi")
@Composable
private fun agentLogoRes(assistant: String): Int? {
    val name = when (assistant.lowercase()) {
        "claude"   -> "claude_mark"
        "codex"    -> "codex_mark"
        "gemini"   -> "gemini_mark"
        "opencode" -> "opencode_mark"
        else       -> return null
    }
    val ctx = LocalContext.current
    return ctx.resources.getIdentifier(name, "drawable", ctx.packageName).takeIf { it != 0 }
}

/**
 * Per-agent optical-size correction for the bundled brand-mark drawables.
 * The source art isn't uniformly cropped -- gemini's and opencode's marks
 * carry more effective padding than claude's/codex's at the same
 * `ContentScale.Fit` frame, so they read visibly smaller even in an
 * identical container (device feedback). Applied as an extra multiplier by
 * both [AgentAvatar] and [AgentGlyph] so every agent's logo occupies
 * roughly the same optical weight at any size. `1f` is a no-op; only the
 * two under-sized marks are scaled up. Mirror of iOS
 * `AgentAvatar.glyphRenderScale(forAgent:)`.
 */
private fun agentGlyphRenderScale(assistant: String): Float =
    when (assistant.lowercase()) {
        "gemini", "opencode" -> 1.25f
        else -> 1f
    }

/**
 * Per-agent brand glyph. Claude → a sparkle, Codex → the code-brackets
 * mark. Returns null for agents we don't have a glyph for (they fall
 * back to [monogramFor]). Neutral Material symbols rather than shipping
 * Anthropic / OpenAI logo artwork in the APK. The string key is exposed
 * via [agentGlyphKey] for plain-JVM unit tests.
 */
private fun agentGlyph(assistant: String): ImageVector? = when (agentGlyphKey(assistant)) {
    "sparkle" -> Icons.Filled.AutoAwesome
    "code"    -> Icons.Filled.Code
    else      -> null
}

internal fun agentGlyphKey(assistant: String): String? = when (assistant.lowercase()) {
    "claude" -> "sparkle"
    "codex"  -> "code"
    else     -> null
}

/**
 * Per-agent monogram. Codex breaks the "first letter" pattern — "C"
 * already belongs to Claude, so Codex gets "X" (Codex eXecution, and
 * visually distinct from C). Everything else is the first letter.
 */
internal fun monogramFor(assistant: String): String = when (assistant.lowercase()) {
    "claude"   -> "C"
    "codex"    -> "X"
    "hermes"   -> "H"
    "pi"       -> "π"
    "opencode" -> "O"
    else       -> assistant.take(1).uppercase()
}

/** Visible-for-tests overload that bypasses Compose for unit tests. */
internal fun agentAvatarMonogram(assistant: String): String = monogramFor(assistant)

File: apps/ios/Sources/Shared/AgentAvatar.swift
import SwiftUI
import UIKit

/// Small circular avatar for an agent (claude, codex, hermes, pi,
/// opencode). Used in any place that lists or picks agents — the
/// `AgentPickerSheet` rows, the `ThreadSwitcherSheet` peek strip and
/// row list, and the `SessionInfoView` hero. Not used inside the
/// chat composer or the header pill — those are already tinted via
/// `ConduitTheme.accent(forAgent:)` directly.
///
/// Renders the real brand mark (or an SF Symbol / monogram fallback)
/// tinted in the agent's theme color over a tint-at-18%-opacity disc,
/// with a solid-tint ring -- the `AgentDot` idiom (see
/// `ConduitUI.AgentDot` in `ConduitFlowAtoms.swift`), not a white disc.
/// A white disc read as a jarring light patch against the dark sheet
/// and buried the mark's own color identity (device feedback); tinting
/// the mark itself (`.renderingMode(.template)`, via `AgentGlyph`) keeps
/// the dark Codex knot visible without one.
struct AgentAvatar: View {
    let assistant: String
    var size: CGFloat = 24
    @Environment(\.neonTheme) private var neon

    var body: some View {
        let tint = neon.agentTint(forAgent: assistant)
        return ZStack {
            Circle().fill(tint.opacity(0.18))
            AgentGlyph(assistant: assistant, size: size * 0.6)
        }
        .frame(width: size, height: size)
        .clipShape(Circle())
        .overlay(
            Circle().strokeBorder(tint, lineWidth: 1.5)
        )
        .accessibilityLabel(Text(assistant.capitalized))
    }

    /// Real brand-logo asset name for an agent, if the app owner has
    /// bundled the official artwork. Returns nil when no asset is present
    /// in the catalog (looked up at runtime) — the avatar then degrades to
    /// [symbol] / [monogram], so a missing asset never breaks the build.
    /// The artwork itself is supplied by the app owner under the trademark
    /// attribution shipped in the Licenses screen — we don't bundle it here.
    static func logoAsset(forAgent assistant: String) -> String? {
        let name: String
        switch assistant.lowercased() {
        case "claude":    name = "ClaudeMark"
        case "codex":     name = "CodexMark"
        case "gemini":    name = "GeminiMark"
        case "opencode":  name = "OpencodeMark"
        default:          return nil
        }
        return UIImage(named: name) != nil ? name : nil
    }

    /// Per-agent optical-size correction for the bundled brand-mark PNGs.
    /// The source art isn't uniformly cropped -- gemini's and opencode's
    /// marks carry more effective padding than claude's/codex's at the
    /// same `scaledToFit` frame, so they read visibly smaller even in an
    /// identical container (device feedback). Applied as an extra
    /// multiplier by both [AgentAvatar] and [AgentGlyph] so every agent's
    /// logo occupies roughly the same optical weight at any size. `1.0` is
    /// a no-op; only the two under-sized marks are scaled up.
    static func glyphRenderScale(forAgent assistant: String) -> CGFloat {
        switch assistant.lowercased() {
        case "gemini", "opencode": return 1.25
        default: return 1.0
        }
    }

    /// Per-agent brand glyph as an SF Symbol name. Claude → a sparkle,
    /// Codex → the code-brackets mark. Returns nil for agents we don't
    /// have a glyph for (they fall back to [monogram]). We use neutral
    /// system symbols rather than shipping Anthropic / OpenAI logo
    /// artwork in the bundle.
    static func symbol(forAgent assistant: String) -> String? {
        switch assistant.lowercased() {
        case "claude": return "sparkle"
        case "codex":  return "chevron.left.forwardslash.chevron.right"
        default:       return nil
        }
    }

    /// Static form of the per-agent monogram, so other bare-glyph
    /// components (e.g. [AgentGlyph]) can reuse it without instantiating
    /// a full `AgentAvatar`.
    static func monogram(forAgent assistant: String) -> String {
        switch assistant.lowercased() {
        case "claude":   return "C"
        case "codex":    return "X"
        case "hermes":   return "H"
        case "pi":       return "π"
        case "opencode": return "O"
        default:
            return String(assistant.prefix(1)).uppercased()
        }
    }
}

/// Bare tinted per-agent glyph -- no background disc/tile, just the real
/// brand mark (Claude starburst / Codex knot / opencode mark), or an SF
/// Symbol / monogram letter fallback when no mark is bundled, tinted with
/// the agent's theme color. This is the ConduitUI list-row idiom ("rows
/// lead with a bare tinted symbol, not a filled tile" -- see
/// `ConduitListRow`). Use this instead of `AgentAvatar` in any row-leading
/// position; reserve `AgentAvatar`'s filled disc for picker / hero contexts
/// that want a heavier visual anchor.
struct AgentGlyph: View {
    let assistant: String
    var size: CGFloat = 20
    @Environment(\.neonTheme) private var neon

    var body: some View {
        let tint = neon.agentTint(forAgent: assistant)
        return Group {
            if let asset = AgentAvatar.logoAsset(forAgent: assistant) {
                // Real brand mark, rendered as a template (alpha-mask) image
                // so it tints exactly like the SF Symbol fallback below --
                // bare mark, no white tile/disc behind it. Scaled by the
                // same optical-size correction as `AgentAvatar` (see
                // `glyphRenderScale`) so gemini/opencode don't read smaller
                // than claude/codex; the resulting frame stays well inside
                // the outer `size` container, so this can't overflow/clip.
                let glyphSize = size * 0.62 * AgentAvatar.glyphRenderScale(forAgent: assistant)
                Image(asset)
                    .renderingMode(.template)
                    .resizable()
                    .scaledToFit()
                    .frame(width: glyphSize, height: glyphSize)
            } else if let symbol = AgentAvatar.symbol(forAgent: assistant) {
                Image(systemName: symbol)
                    .font(.system(size: size * 0.62, weight: .semibold))
            } else {
                Text(AgentAvatar.monogram(forAgent: assistant))
                    .font(.system(size: size * 0.58, weight: .heavy, design: .rounded))
            }
        }
        .foregroundStyle(tint)
        .frame(width: size, height: size)
        .accessibilityLabel(Text(assistant.capitalized))
    }
}

#Preview("Agent avatars") {
    HStack(spacing: 12) {
        AgentAvatar(assistant: "claude")
        AgentAvatar(assistant: "codex")
        AgentAvatar(assistant: "hermes")
        AgentAvatar(assistant: "pi")
        AgentAvatar(assistant: "opencode")
        AgentAvatar(assistant: "unknown")
    }
    .padding()
}
